# 4.5: Data Consistency Checks

Content List:
1. Import Libraries, Create Path and Import Datasets
2. Create a test df to explore mixed datatypes
3. Find and Handle Missing Values
4. Identify and Handle Duplicate Values
5. Export Updated Product df

Assignment 4.5 Begin

6. Explore the Orders df
7. Check for mixed type data
8. Check for and investigate missing values
9. Check for duplicate values
10. Export updated Orders df

In [2]:
#Importing Libraries
import pandas as pd
import numpy as nb
import os

In [3]:
#Create Path
path = r'C:\Users\ryanc\OneDrive\Documents\Data Analytics Exercises\Python\Instacart Basket Analysis'

In [4]:
#Import Order df
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [5]:
#Import Product df
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

2. Create a test df to explore mixed datatypes

In [7]:
#Create a test dataframe
df_test = pd.DataFrame()

In [9]:
df_test['mix'] = ['a','b',1,True]

In [10]:
df_test

mix
0     a
1     b
2     1
3  True

In [11]:
#Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


3. Find and Handle Missing Values

In [12]:
#Identify how many missing values exist
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
#View missing values from product name column only
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
#Drop missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

4. Identify and Handle Duplicate Values

In [17]:
#Create new dataframe of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
#View Duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
#Drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

5. Export Updated Product df

In [20]:
#Export Cleaned Data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

Assignment 4.5 Begin

6. Explore the Orders df

Q2. Share in a markdown cell whether anything about the data looks off or should be investigated further

In [21]:
#Describe the Order df
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Q2: 
Nothing about this data initially looks off to me. days since prior order ranges from 1 to 30 days, order number ranges from 1 to 100. Therefor I assume the data stop counting days since prior order at a max of 30 and either no customer has ordered more than 100 times or they also stop counting order numbers at 100. Order_dow ranges from 0-6 for 7 days in a week, order_hour ranges from 0-23 for 24 hours in a day. order_id max equal the order_id count as it should.

7. Check for mixed type data

In [23]:
#Check for mixed type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)


No mix type data exists

8. Check for and investigate missing values

In [28]:
#Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Question 5:
The only column with missing values is days_since_prior_order. Further investigating is needed.

In [29]:
#View missing values from days_since_prior_order column only
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [30]:
#View missing values df
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

The number of missing values from the column 'days_since_prior_order' is 206209, this perfectly matches the number of user_id's
there are in total. And each missing value is correlated with an order_number of '1'. Therefore this value is not really missing
as anyone who is placing their first order would not have any 'days_since_prior_order'. No further action is necessary.

9. Check for duplicate values

In [26]:
#Check for duplicates
df_dups = df_ords[df_ords.duplicated()]

In [27]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicate values found. No further action is necessary.

10. Export updated Orders df

In [31]:
#Export Cleaned Order Data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

#Note that cleaned Product Data was already exported above